crates/bevy_diagnostic/src/log_diagnostics_plugin.rs

```diff
@@ -2,103 +2,102 @@ use super::{Diagnostic, DiagnosticId, Diagnostics};
-pub struct PrintDiagnosticsState {
+/// State used by the [LogDiagnosticsPlugin]
+pub struct LogDiagnosticsState {
     timer: Timer,
     filter: Option<Vec<DiagnosticId>>,
 }
 
-impl Default for PrintDiagnosticsPlugin {
+impl Default for LogDiagnosticsPlugin {
     fn default() -> Self {
-        PrintDiagnosticsPlugin {
+        LogDiagnosticsPlugin {
             debug: false,
             wait_duration: Duration::from_secs(1),
             filter: None,
         }
     }
 }
 
-impl Plugin for PrintDiagnosticsPlugin {
+impl Plugin for LogDiagnosticsPlugin {
     fn build(&self, app: &mut bevy_app::AppBuilder) {
-        app.add_resource(PrintDiagnosticsState {
+        app.add_resource(LogDiagnosticsState {
             timer: Timer::new(self.wait_duration, true),
             filter: self.filter.clone(),
         });
 
         if self.debug {
             app.add_system_to_stage(
                 stage::POST_UPDATE,
-                Self::print_diagnostics_debug_system.system(),
+                Self::log_diagnostics_debug_system.system(),
             );
         } else {
-            app.add_system_to_stage(stage::POST_UPDATE, Self::print_diagnostics_system.system());
+            app.add_system_to_stage(stage::POST_UPDATE, Self::log_diagnostics_system.system());
         }
     }
 }
 
-impl PrintDiagnosticsPlugin {
+impl LogDiagnosticsPlugin {
     pub fn filtered(filter: Vec<DiagnosticId>) -> Self {
-        PrintDiagnosticsPlugin {
+        LogDiagnosticsPlugin {
             filter: Some(filter),
             ..Default::default()
         }
     }
 
-    fn print_diagnostic(diagnostic: &Diagnostic) {
+    fn log_diagnostic(diagnostic: &Diagnostic) {
         if let Some(value) = diagnostic.value() {
-            print!("{:<65}: {:<10.6}", diagnostic.name, value);
             if let Some(average) = diagnostic.average() {
-                print!("  (avg {:.6})", average);
+                info!(
+                    "{:<65}: {:<10.6}  (avg {:.6})",
+                    diagnostic.name, value, average
+                );
+            } else {
+                info!("{:<65}: {:<10.6}", diagnostic.name, value);
             }
-
-            println!("\n");
         }
     }
 
-    pub fn print_diagnostics_system(
-        mut state: ResMut<PrintDiagnosticsState>,
+    pub fn log_diagnostics_system(
+        mut state: ResMut<LogDiagnosticsState>,
         time: Res<Time>,
         diagnostics: Res<Diagnostics>,
     ) {
         if state.timer.tick(time.delta_seconds()).finished() {
-            println!("Diagnostics:");
-            println!("{}", "-".repeat(93));
             if let Some(ref filter) = state.filter {
                 for diagnostic in filter.iter().map(|id| diagnostics.get(*id).unwrap()) {
-                    Self::print_diagnostic(diagnostic);
+                    Self::log_diagnostic(diagnostic);
                 }
             } else {
                 for diagnostic in diagnostics.iter() {
```

Comment:
Can these be sorted so that the diagnostics are in a consistent order?

It's not helpful every time you restart the app getting a different order.

Reply:
Currently you can have them ordered if you use a filter, as the filter is stored in a `Vec` and then iterated over.

Without filter, the diagnostics are stored in a `HashMap` and that doesn't keep order... We could switch to a [`BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html), or add a `Vec` to keep order of insertion and use it to iterate... What do you think?

Reply:
So `BTreeMap` would probably be a good start, although a `Vec` sorted by `name` or by insertion order could both also be a good choice.

Basically, anything deterministic would be better than the current situation.

Reply:
Moved to a `BTreeMap`. After looking a little into a `Vec` solution, I didn't like it as it would have mean every diagnostic recording a value would have to go through the `Vec`...

Reply:
O(logn) is better than O(n), but for a large number of diagnostics idk if its a good idea to replace a hash map. While we currently only have a small number, I don't want to assume that will always be the case. Maybe LogDiagnosticsPlugin could store its own order? Sorting is a behavior inherent to the LogDiagnosticsPlugin so that makes sense to me.

The issue is keeping the order synced up / accounting for new diagnostics. Lol we could have a Diagnostic event. Ex: `DiagnosticEvent::Added(DiagnosticId)`, `DiagnosticEvent::Removed(DiagnosticId)`, etc

Reply:
I looked into adding an event, seems not easy as adding a diagnostics can happen at app build time by mutating a resource, so the new diagnostic plugin would be responsible from calling the event (I think)
Keeping the order in LogDiagnosticsPlugin without events seemed not pretty
I added a `BTreeSet` of the `DiagnosticId` to `Diagnostics` that is updated when adding a diagnostics, and that is used for iterating orderedly